beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go

```diff
@@ -714,3 +714,94 @@ func (f *ForkChoice) ParentRoot(root [32]byte) ([32]byte, error) {
 	}
 	return n.parent.root, nil
 }
+
+// UpdateVotesOnPayloadAttestation processes a new aggregated
+// payload attestation message and updates
+// the Payload Timeliness Committee (PTC) votes for the corresponding block.
+// It may update payload boost roots based on the attestation data.
+func (s *Store) UpdateVotesOnPayloadAttestation(
+	payloadAttestation *ethpb.PayloadAttestation,
+	isFromBlock bool) error {
+	// Extract the attestation data and convert the beacon block root to a 32-byte array
+	data := payloadAttestation.Data
+	blockRoot := bytesutil.ToBytes32(data.BeaconBlockRoot)
+
+	// Check if the block exists in the store
+	node, ok := s.nodeByRoot[blockRoot]
+	if !ok {
+		return errors.New("beacon block root unknown")
+	}
+
+	// Only update payload boosts with attestations from a block if the block is for the current slot and it's early
+	if isFromBlock {
+		currentSlot := slots.CurrentSlot(s.genesisTime)
+		// Only process if the attestation is for the previous slot
+		if data.Slot+1 != currentSlot {
+			return nil
+		}
+		// Only process if we're still in the early part of the slot
+		// This check ensures we only boost timely attestations
+		timeIntoSlot, err := slots.SecondsSinceSlotStart(currentSlot, s.genesisTime, uint64(time.Now().Unix()))
+		if err != nil {
+			log.WithError(err).Error("could not compute seconds since slot start")
+		}
+		if timeIntoSlot >= params.BeaconConfig().SecondsPerSlot/params.BeaconConfig().IntervalsPerSlot {
+			return nil
+		}
+	}
+
+	// Initialize the PTC vote for this node if it doesn't exist
+	// The size is set to PTCSize, which is a preset constant
+	if node.ptcVote == nil {
+		node.ptcVote = make([]primitives.PTCStatus, fieldparams.PTCSize)
+	}
+
+	// Update the PTC votes based on the attestation
+	// We iterate through each bit in the AggregationBits
+	// If a bit is not already set, we update the corresponding PTC vote
+	for i := uint64(0); i < fieldparams.PTCSize; i++ {
+		if !payloadAttestation.AggregationBits.BitAt(i) {
+			node.ptcVote[i] = data.PayloadStatus
+		}
+	}
+
+	// Only update payload boosts if they haven't been applied to this node yet
+	// This check prevents unnecessary computations
+	if s.payloadRevealBoostRoot != node.root && s.payloadWithholdBoostRoot != node.root {
+		// Update payload boosts if necessary
+		s.updatePayloadBoosts(node)
+	}
+
+	return nil
+}
+
+// updatePayloadBoosts checks the PTC votes for a given node and updates
+// the payload reveal and withhold boost roots if the necessary thresholds are met.
+func (s *Store) updatePayloadBoosts(node *Node) {
+	presentCount := 0
+	withheldCount := 0
+
+	// Count the number of PRESENT and WITHHELD votes
+	for _, vote := range node.ptcVote {
+		if vote == primitives.PAYLOAD_PRESENT {
+			presentCount++
+		} else if vote == primitives.PAYLOAD_ABSENT {
+			withheldCount++
```

Comment:
This should check for `PAYLOAD_WITHHELD` instead

Reply:
thanks! fixed this in the next commit